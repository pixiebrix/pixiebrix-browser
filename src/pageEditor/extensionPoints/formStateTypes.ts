/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type NormalizedAvailability } from "@/blocks/types";
import { type ElementInfo } from "@/pageScript/frameworks";
import {
  type ContextMenuConfig,
  type ContextMenuTargetMode,
  type MenuDefaultOptions as ContextMenuDefaultOptions,
} from "@/extensionPoints/contextMenu";
import {
  type MenuItemExtensionConfig,
  type MenuPosition,
} from "@/extensionPoints/menuItemExtension";
import { type PanelConfig } from "@/extensionPoints/panelExtension";
import {
  type QuickBarConfig,
  type QuickBarDefaultOptions,
  type QuickBarTargetMode,
} from "@/extensionPoints/quickBarExtension";
import {
  type SidebarConfig,
  type Trigger as SidebarTrigger,
} from "@/extensionPoints/sidebarExtension";
import {
  type AttachMode,
  type ReportMode,
  type TargetMode,
  type Trigger as TriggerTrigger,
} from "@/extensionPoints/triggerExtension";
import {
  type CustomEventOptions,
  type DebounceOptions,
  type ExtensionPointType,
} from "@/extensionPoints/types";
import { type Except } from "type-fest";
import { type Menus } from "webextension-polyfill";
import {
  type BaseExtensionState,
  type BaseExtensionPointState,
  type SingleLayerReaderConfig,
  type BaseFormState,
} from "./elementConfig";

// ActionFormState
type ActionExtensionState = BaseExtensionState &
  Except<MenuItemExtensionConfig, "action">;
type ActionExtensionPointState = BaseExtensionPointState & {
  definition: {
    type: ExtensionPointType;
    containerSelector: string;
    position?: MenuPosition;
    template: string;
    /**
     * @since 1.7.16
     */
    targetMode?: "document" | "eventTarget";
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
  traits?: {
    style: {
      mode: "default" | "inherit";
    };
  };
};

export interface ActionFormState
  extends BaseFormState<ActionExtensionState, ActionExtensionPointState> {
  type: "menuItem";
  containerInfo: ElementInfo;
}

// SidebarFormState
type SidebarExtensionState = BaseExtensionState & Except<SidebarConfig, "body">;

export type SidebarExtensionPointState = BaseExtensionPointState & {
  definition: BaseExtensionPointState["definition"] & {
    /**
     * Sidebar trigger (default="load")
     * @since 1.6.5
     */
    trigger: SidebarTrigger;

    /**
     * Debouncing props
     * @since 1.6.5
     */
    debounce: DebounceOptions | null;

    /**
     * Custom trigger props
     * @since 1.6.5
     */
    customEvent: CustomEventOptions | null;
  };
};

export interface SidebarFormState
  extends BaseFormState<SidebarExtensionState, SidebarExtensionPointState> {
  type: "actionPanel";
}

// TriggerFormState
export type TriggerExtensionPointState = BaseExtensionPointState & {
  definition: {
    type: ExtensionPointType;
    rootSelector: string | null;
    trigger: TriggerTrigger;
    reader: SingleLayerReaderConfig;
    attachMode: AttachMode;
    targetMode: TargetMode;
    reportMode: ReportMode;

    isAvailable: NormalizedAvailability;

    // Debouncing props
    debounce: DebounceOptions;

    // Custom tigger props
    customEvent: CustomEventOptions;

    // Interval props
    intervalMillis: number | null;
    background: boolean | null;
  };
};

export function isTriggerExtensionPoint(
  extensionPoint: BaseExtensionPointState
): extensionPoint is TriggerExtensionPointState {
  return extensionPoint.definition.type === "trigger";
}

export interface TriggerFormState
  extends BaseFormState<BaseExtensionState, TriggerExtensionPointState> {
  type: "trigger";
}

// PanelFormState
export type PanelTraits = {
  style: {
    mode: "default" | "inherit";
  };
};

type PanelExtensionState = BaseExtensionState & Except<PanelConfig, "body">;
type PanelExtensionPointState = BaseExtensionPointState & {
  definition: {
    type: ExtensionPointType;
    containerSelector: string;
    position?: MenuPosition;
    template: string;
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
  traits: PanelTraits;
};

export interface PanelFormState
  extends BaseFormState<PanelExtensionState, PanelExtensionPointState> {
  type: "panel";

  containerInfo: ElementInfo;
}

// ContextMenuFormState
type ContextMenuExtensionState = BaseExtensionState &
  Except<ContextMenuConfig, "action">;
type ContextMenuExtensionPointState = BaseExtensionPointState & {
  definition: {
    type: ExtensionPointType;
    defaultOptions: ContextMenuDefaultOptions;
    documentUrlPatterns: string[];
    contexts: Menus.ContextType[];
    targetMode: ContextMenuTargetMode;
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
};

export interface ContextMenuFormState
  extends BaseFormState<
    ContextMenuExtensionState,
    ContextMenuExtensionPointState
  > {
  type: "contextMenu";
}

// QuickBarFormState
type QuickBarExtensionState = BaseExtensionState &
  Except<QuickBarConfig, "action">;
type QuickBarExtensionPointState = BaseExtensionPointState & {
  definition: {
    type: ExtensionPointType;
    defaultOptions: QuickBarDefaultOptions;
    documentUrlPatterns: string[];
    contexts: Menus.ContextType[];
    targetMode: QuickBarTargetMode;
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
};

export function isQuickBarExtensionPoint(
  extensionPoint: BaseExtensionPointState
): extensionPoint is QuickBarExtensionPointState {
  return extensionPoint.definition.type === "quickBar";
}

export interface QuickBarFormState
  extends BaseFormState<QuickBarExtensionState, QuickBarExtensionPointState> {
  type: "quickBar";
}

export type FormState =
  | ActionFormState
  | SidebarFormState
  | TriggerFormState
  | PanelFormState
  | ContextMenuFormState
  | QuickBarFormState;
