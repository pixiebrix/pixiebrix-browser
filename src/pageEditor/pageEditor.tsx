/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/bootstrapWithoutRem.css";
import "@/utils/global.scss";
import "@/utils/layout.scss";

import "@/extensionContext";
import "@/development/darkMode.js";

import { initMessengerLogging } from "@/development/messengerLogging";

import ReactDOM from "react-dom";
import React from "react";
import Panel from "@/pageEditor/Panel";
import { watchNavigation } from "@/pageEditor/protocol";
import { initToaster } from "@/utils/notify";
import { initRuntimeLogging } from "@/development/runtimeLogging";
import { initPerformanceMonitoring } from "@/telemetry/performance";
import { setPlatform } from "@/platform/platformContext";
import extensionPagePlatform from "@/extensionPages/extensionPagePlatform";

async function init() {
  setPlatform(extensionPagePlatform);
  void initMessengerLogging();
  void initRuntimeLogging();
  try {
    await initPerformanceMonitoring();
  } catch (error) {
    console.error("Failed to initialize performance monitoring", error);
  }

  watchNavigation();
  initToaster();

  ReactDOM.render(<Panel />, document.querySelector("#container"));
}

void init();
