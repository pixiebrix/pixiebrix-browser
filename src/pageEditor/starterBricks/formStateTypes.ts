/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type ElementInfo } from "@/utils/inference/selectorTypes";
import {
  type ContextMenuConfig,
  type ContextMenuTargetMode,
  type MenuDefaultOptions as ContextMenuDefaultOptions,
} from "@/starterBricks/contextMenu/contextMenuTypes";
import {
  type ButtonStarterBrickConfig,
  type ButtonPosition,
} from "@/starterBricks/button/buttonStarterBrickTypes";
import {
  type QuickBarConfig,
  type QuickBarDefaultOptions,
  type QuickBarTargetMode,
} from "@/starterBricks/quickBar/quickBarTypes";
import {
  type SidebarConfig,
  type Trigger as SidebarTrigger,
} from "@/starterBricks/sidebar/sidebarStarterBrickTypes";
import {
  type AttachMode,
  type ReportMode,
  type TargetMode,
  type Trigger as TriggerTrigger,
} from "@/starterBricks/trigger/triggerStarterBrickTypes";
import {
  type CustomEventOptions,
  type DebounceOptions,
} from "@/starterBricks/types";
import {
  type StarterBrickType,
  type StarterBrickTypes,
} from "@/types/starterBrickTypes";
import { type Except } from "type-fest";
import { type Menus } from "webextension-polyfill";
import {
  type QuickBarProviderConfig,
  type QuickBarProviderDefaultOptions,
} from "@/starterBricks/quickBarProvider/quickBarProviderTypes";
import {
  type BaseStarterBrickState,
  type BaseModComponentState,
  type BaseFormState,
  type SingleLayerReaderConfig,
} from "@/pageEditor/baseFormStateTypes";
import { type NormalizedAvailability } from "@/types/availabilityTypes";
import { type Nullishable } from "@/utils/nullishUtils";

// ButtonFormState
type ButtonModComponentState = BaseModComponentState &
  Except<ButtonStarterBrickConfig, "action">;
type ButtonStarterBrickState = BaseStarterBrickState & {
  definition: {
    type: StarterBrickType;
    containerSelector: string;
    position?: ButtonPosition;
    template: string;
    /**
     * @since 1.7.16
     */
    targetMode?: "document" | "eventTarget";
    /**
     * @since 1.7.28
     */
    attachMode?: "once" | "watch";
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
  traits?: {
    style: {
      mode: "default" | "inherit";
    };
  };
};

export interface ButtonFormState
  extends BaseFormState<ButtonModComponentState, ButtonStarterBrickState> {
  type: typeof StarterBrickTypes.BUTTON;
  containerInfo: ElementInfo | null;
}

// SidebarFormState
type SidebarModComponentState = BaseModComponentState &
  Except<SidebarConfig, "body">;

type SidebarStarterBrickState = BaseStarterBrickState & {
  definition: BaseStarterBrickState["definition"] & {
    /**
     * Sidebar trigger (default="load")
     * @since 1.6.5
     */
    trigger: SidebarTrigger;

    /**
     * Debouncing props
     * @since 1.6.5
     */
    debounce: Nullishable<DebounceOptions>;

    /**
     * Custom trigger props
     * @since 1.6.5
     */
    customEvent: Nullishable<CustomEventOptions>;
  };
};

export interface SidebarFormState
  extends BaseFormState<SidebarModComponentState, SidebarStarterBrickState> {
  type: "actionPanel";
}

// TriggerFormState
type TriggerStarterBrickState = BaseStarterBrickState & {
  definition: {
    type: StarterBrickType;
    rootSelector?: string;
    trigger?: TriggerTrigger;
    reader: SingleLayerReaderConfig;
    attachMode?: AttachMode;
    targetMode?: TargetMode;
    reportMode?: ReportMode;
    /**
     * @since 1.7.34
     */
    showErrors?: boolean;

    isAvailable: NormalizedAvailability;

    // Debouncing props
    debounce?: DebounceOptions;

    // Custom tigger props
    customEvent?: CustomEventOptions;

    // Interval props
    intervalMillis?: number;
    background?: boolean;
  };
};

export function isTriggerStarterBrick(
  starterBrick: BaseStarterBrickState,
): starterBrick is TriggerStarterBrickState {
  return starterBrick.definition.type === "trigger";
}

export interface TriggerFormState
  extends BaseFormState<BaseModComponentState, TriggerStarterBrickState> {
  type: "trigger";
}

// ContextMenuFormState
type ContextMenuModComponentState = BaseModComponentState &
  Except<ContextMenuConfig, "action">;
type ContextMenuStarterBrickState = BaseStarterBrickState & {
  definition: {
    type: StarterBrickType;
    defaultOptions: ContextMenuDefaultOptions;
    documentUrlPatterns: string[];
    contexts: Menus.ContextType[];
    targetMode: ContextMenuTargetMode;
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
};

export interface ContextMenuFormState
  extends BaseFormState<
    ContextMenuModComponentState,
    ContextMenuStarterBrickState
  > {
  type: "contextMenu";
}

// QuickBarFormState
type QuickBarModComponentState = BaseModComponentState &
  Except<QuickBarConfig, "action">;
type QuickBarStarterBrickState = BaseStarterBrickState & {
  definition: {
    type: StarterBrickType;
    defaultOptions: QuickBarDefaultOptions;
    documentUrlPatterns: string[];
    contexts: Menus.ContextType[];
    targetMode: QuickBarTargetMode;
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
};

// QuickBarFormState
type QuickBarProviderModComponentState = BaseModComponentState &
  Except<QuickBarProviderConfig, "generator"> & {
    rootAction?: QuickBarProviderConfig["rootAction"];
  };
type QuickBarProviderStarterBrickState = BaseStarterBrickState & {
  definition: {
    type: StarterBrickType;
    defaultOptions: QuickBarProviderDefaultOptions;
    documentUrlPatterns: string[];
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
};

export function isQuickBarStarterBrick(
  starterBrick: BaseStarterBrickState,
): starterBrick is QuickBarStarterBrickState {
  return ["quickBar", "quickBarProvider"].includes(
    starterBrick.definition.type,
  );
}

export interface QuickBarFormState
  extends BaseFormState<QuickBarModComponentState, QuickBarStarterBrickState> {
  type: "quickBar";
}

export interface QuickBarProviderFormState
  extends BaseFormState<
    QuickBarProviderModComponentState,
    QuickBarProviderStarterBrickState
  > {
  type: "quickBarProvider";
}

/**
 * @deprecated We want to deconstruct ComponentFormState and using reducers instead of
 * useEffect/useAsyncEffect for defaulting, cleaning up integration configurations, etc.
 */
export type ModComponentFormState =
  | ButtonFormState
  | TriggerFormState
  | SidebarFormState
  | ContextMenuFormState
  | QuickBarFormState
  | QuickBarProviderFormState;
