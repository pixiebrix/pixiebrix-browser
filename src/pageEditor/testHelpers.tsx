/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import { appApi } from "@/services/api";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
