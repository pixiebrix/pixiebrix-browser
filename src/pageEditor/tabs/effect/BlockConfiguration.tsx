/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useEffect, useMemo, useRef } from "react";
import { type RegistryId } from "@/core";
import { getIn, useField, useFormikContext } from "formik";
import useBlockOptions from "@/hooks/useBlockOptions";
import { Card } from "react-bootstrap";
import SchemaFieldContext from "@/components/fields/schemaFields/SchemaFieldContext";
import devtoolFieldOverrides from "@/pageEditor/fields/devtoolFieldOverrides";
import Loader from "@/components/Loader";
import ConnectedFieldTemplate from "@/components/form/ConnectedFieldTemplate";
import { joinName } from "@/utils";
import { useAsyncState } from "@/hooks/common";
import SelectWidget, {
  type Option,
} from "@/components/form/widgets/SelectWidget";
import { partial } from "lodash";
import { type BlockConfig, type BlockWindow } from "@/blocks/types";
import AdvancedLinks, {
  DEFAULT_WINDOW_VALUE,
} from "@/pageEditor/tabs/effect/AdvancedLinks";
import { type SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import SchemaField from "@/components/fields/schemaFields/SchemaField";
import FieldSection from "@/pageEditor/fields/FieldSection";
import getType from "@/runtime/getType";
import { type FormState } from "@/pageEditor/extensionPoints/formStateTypes";
import ConfigurationTitle from "./ConfigurationTitle";
import { inputProperties } from "@/helpers";

const rootModeOptions = [
  { label: "Document", value: "document" },
  { label: "Element", value: "element" },
  { label: "Inherit", value: "inherit" },
];

const targetOptions: Array<Option<BlockWindow>> = [
  { label: "Current Tab (self)", value: "self" },
  { label: "Opener Tab (opener)", value: "opener" },
  { label: "Target Tab (target)", value: "target" },
  { label: "Top-level Frame (top)", value: "top" },
  { label: "All Tabs (broadcast)", value: "broadcast" },
];

const BlockConfiguration: React.FunctionComponent<{
  name: string;
  blockId: RegistryId;
}> = ({ name, blockId }) => {
  const configName = partial(joinName, name);

  const context = useFormikContext<FormState>();
  const [config] = useField<BlockConfig>(name);
  const [_rootField, _rootFieldMeta, rootFieldHelpers] = useField<BlockConfig>(
    configName("root")
  );
  const blockErrors = getIn(context.errors, name);

  const [{ block, error }, BlockOptions] = useBlockOptions(blockId);

  // Conditionally show Advanced options "Condition" and "Target" depending on the value of blockType.
  // If blockType is undefined, don't show the options.
  // If error happens, behavior is undefined.
  const [blockType] = useAsyncState(async () => getType(block), [block]);

  const [isRootAware] = useAsyncState(async () => {
    const inputSchema = inputProperties(block.inputSchema);
    // Handle DOM bricks that were upgraded to be root-aware
    if ("isRootAware" in inputSchema) {
      return Boolean(config.value.config.isRootAware);
    }

    return block.isRootAware();
  }, [block, config.value.config.isRootAware]);

  const advancedOptionsRef = useRef<HTMLDivElement>();

  useEffect(
    () => {
      // Effect to clear out unused `root` field. Technically, `root` could contain a selector when used with `document`
      // or `inherit` mode, but we don't want to support that in the Page Editor because it's legacy behavior.
      if (config.value.rootMode !== "element") {
        rootFieldHelpers.setValue(null);
      }
    }, // eslint-disable-next-line react-hooks/exhaustive-deps -- rootFieldHelpers changes reference every render
    [config.value.rootMode]
  );

  const rootElementSchema: SchemaFieldProps = useMemo(
    () => ({
      name: configName("root"),
      label: "Target Element",
      description: (
        <span>
          The target element for the brick. Provide element reference{" "}
          <code>@input.element.ref</code>, or a reference generated with the
          Element Reader, For-Each Element, or Traverse Elements brick
        </span>
      ),
      // If the field is visible, it's required
      isRequired: true,
      schema: {
        // The type is https://app.pixiebrix.com/schemas/element#, but schema field doesn't support that
        // XXX: this should really restrict to "variable" entry. Text values will also be interpreted properly, it's
        //  there's just no use-case for hard coded element uuids or text template expressions
        type: "string",
      },
    }),
    [configName]
  );

  const ifSchemaProps: SchemaFieldProps = useMemo(
    () => ({
      name: configName("if"),
      schema: {
        type: ["string", "number", "boolean"],
      },
      label: "Condition",
      description: (
        <p>
          Condition determining whether or not to execute the brick. Truthy
          string values are&nbsp;
          <code>true</code>, <code>t</code>, <code>yes</code>, <code>y</code>,{" "}
          <code>on</code>, and <code>1</code> (case-insensitive)
        </p>
      ),
    }),
    [configName]
  );

  // Only show if the extension point supports a target mode. menuItem implicitly supports target mode, because
  // it's root-aware if multiple menu items are added to the page.
  // Technically trigger/quickBar/etc. allow the user to pick the target mode. But for now, show the field even if
  // the user has configured the extension point to use the document as the target.
  const showRootMode =
    isRootAware &&
    [
      "trigger",
      "contextMenu",
      "quickBar",
      "quickBarProvider",
      "menuItem",
    ].includes(context.values.type);
  const showIfAndTarget = blockType && blockType !== "renderer";
  const noAdvancedOptions = !showRootMode && !showIfAndTarget;

  return (
    <>
      <AdvancedLinks name={name} scrollToRef={advancedOptionsRef} />

      <Card>
        <FieldSection title={<ConfigurationTitle />}>
          <SchemaFieldContext.Provider value={devtoolFieldOverrides}>
            {blockErrors?.id && (
              <div className="invalid-feedback d-block mb-4">
                Unknown block {blockId}
              </div>
            )}
            {BlockOptions ? (
              <BlockOptions name={name} configKey="config" />
            ) : error ? (
              <div className="invalid-feedback d-block mb-4">{error}</div>
            ) : (
              <Loader />
            )}
          </SchemaFieldContext.Provider>
        </FieldSection>

        <FieldSection title="Advanced Options" bodyRef={advancedOptionsRef}>
          {showIfAndTarget && <SchemaField {...ifSchemaProps} omitIfEmpty />}

          {showRootMode && (
            <ConnectedFieldTemplate
              name={configName("rootMode")}
              label="Target Root Mode"
              as={SelectWidget}
              options={rootModeOptions}
              blankValue="inherit"
              description="The Root Mode controls the page element the brick targets. PixieBrix evaluates selectors relative to the root document/element"
            />
          )}

          {config.value.rootMode === "element" && (
            <SchemaField {...rootElementSchema} omitIfEmpty />
          )}

          {showIfAndTarget && (
            <ConnectedFieldTemplate
              name={configName("window")}
              label="Target Tab/Frame"
              as={SelectWidget}
              options={targetOptions}
              blankValue={DEFAULT_WINDOW_VALUE}
              description="The tab/frame to run the brick. To ensure PixieBrix has permission to run on the tab, add an Extra Permissions pattern that matches the target tab URL"
            />
          )}

          {noAdvancedOptions && (
            <small className="text-muted font-italic">No options to show</small>
          )}
        </FieldSection>
      </Card>
    </>
  );
};

export default BlockConfiguration;
