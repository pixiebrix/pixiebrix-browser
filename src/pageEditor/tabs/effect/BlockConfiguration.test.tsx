/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import BlockConfiguration from "./BlockConfiguration";
import {
  blockConfigFactory,
  blockFactory,
  formStateFactory,
  triggerFormStateFactory,
  quickbarFormStateFactory,
  menuItemFormStateFactory,
  contextMenuFormStateFactory,
  sidebarPanelFormStateFactory,
} from "@/testUtils/factories";
import blockRegistry from "@/blocks/registry";
import { echoBlock } from "@/runtime/pipelineTests/pipelineTestHelpers";
import { screen } from "@testing-library/react";
import { waitForEffect } from "@/testUtils/testHelpers";
import { propertiesToSchema } from "@/validators/generic";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import { render } from "@/pageEditor/testHelpers";
import { type FormState } from "@/pageEditor/extensionPoints/formStateTypes";
import { actions } from "@/pageEditor/slices/editorSlice";

beforeAll(() => {
  registerDefaultWidgets();
  // Precaution
  blockRegistry.clear();
});

afterEach(() => {
  // Being nice to other tests
  blockRegistry.clear();
});

function renderBlockConfiguration(
  element: React.ReactElement,
  initialValues: FormState
) {
  return render(element, {
    initialValues,
    setupRedux(dispatch) {
      dispatch(actions.addElement(initialValues));
      dispatch(actions.selectElement(initialValues.uuid));
      dispatch(
        actions.setElementActiveNodeId(
          initialValues.extension.blockPipeline[0].instanceId
        )
      );
    },
  });
}

test("renders", async () => {
  const block = echoBlock;
  blockRegistry.register([block]);
  const initialState = formStateFactory({ apiVersion: "v3" }, [
    blockConfigFactory({ id: block.id }),
  ]);
  const rendered = renderBlockConfiguration(
    <BlockConfiguration name="extension.blockPipeline[0]" blockId={block.id} />,
    initialState
  );

  await waitForEffect();

  expect(rendered.asFragment()).toMatchSnapshot();
});

describe("shows root mode", () => {
  test.each([
    ["trigger", triggerFormStateFactory],
    ["quickBar", quickbarFormStateFactory],
    ["contextMenu", contextMenuFormStateFactory],
    // `menuItem` must show root mode because root mode is used if the location matches multiple elements on the page
    ["menuItem", menuItemFormStateFactory],
  ])("shows root mode for %s", async (type, factory) => {
    const block = echoBlock;
    blockRegistry.register([block]);
    const initialState = factory({ apiVersion: "v3" }, [
      blockConfigFactory({ id: block.id }),
    ]);
    renderBlockConfiguration(
      <BlockConfiguration
        name="extension.blockPipeline[0]"
        blockId={block.id}
      />,
      initialState
    );

    await waitForEffect();

    const rootModeSelect = screen.getByLabelText("Target Root Mode");

    expect(rootModeSelect).not.toBeNull();
  });

  test("don't show root mode for sidebar panel", async () => {
    const block = echoBlock;
    blockRegistry.register([block]);
    const initialState = sidebarPanelFormStateFactory({ apiVersion: "v3" }, [
      blockConfigFactory({ id: block.id }),
    ]);
    renderBlockConfiguration(
      <BlockConfiguration
        name="extension.blockPipeline[0]"
        blockId={block.id}
      />,
      initialState
    );

    await waitForEffect();

    const rootModeSelect = screen.queryByLabelText("Target Root Mode");

    expect(rootModeSelect).toBeNull();
  });
});

test.each`
  blockName      | propertyName   | expected | readableExpected
  ${"reader"}    | ${"read"}      | ${true}  | ${"should"}
  ${"effect"}    | ${"effect"}    | ${true}  | ${"should"}
  ${"transform"} | ${"transform"} | ${true}  | ${"should"}
  ${"renderer"}  | ${"render"}    | ${false} | ${"should not"}
`(
  "$readableExpected show Condition and Target settings for $blockName",
  async ({ propertyName, expected }) => {
    const block = blockFactory({
      [propertyName]: jest.fn(),
      inputSchema: propertiesToSchema({
        message: {
          type: "string",
        },
      }),
    });

    blockRegistry.register([block]);
    const initialState = triggerFormStateFactory({ apiVersion: "v3" }, [
      blockConfigFactory({ id: block.id }),
    ]);
    renderBlockConfiguration(
      <BlockConfiguration
        name="extension.blockPipeline[0]"
        blockId={block.id}
      />,
      initialState
    );

    await waitForEffect();

    const conditionInput = screen.queryByLabelText("Condition");
    const targetInput = screen.queryByLabelText("Target Tab/Frame");

    if (expected) {
      expect(conditionInput).not.toBeNull();
      expect(targetInput).not.toBeNull();
    } else {
      expect(conditionInput).toBeNull();
      expect(targetInput).toBeNull();
    }
  }
);
