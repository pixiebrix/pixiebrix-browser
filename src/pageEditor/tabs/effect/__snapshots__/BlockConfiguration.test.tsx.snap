// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="formGroup form-group row"
    >
      <div
        class="mb-2 annotation-transition col-12"
      >
        <div
          class="annotationPlaceholder"
        />
      </div>
      <label
        class="label form-label col-form-label col-xl-2 col-lg-3"
        for="extension.blockPipeline[0].config.message"
      >
        message
      </label>
      <div
        class="col-xl-10 col-lg-9"
      >
        <div
          class="root"
        >
          <div
            class="field"
          >
            <textarea
              class="form-control"
              id="extension.blockPipeline[0].config.message"
              name="extension.blockPipeline[0].config.message"
              rows="1"
            />
          </div>
          <div
            class="dropdown dropdown"
            data-test-selected="Text"
            data-testid="toggle-extension.blockPipeline[0].config.message"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle btn btn-link"
              type="button"
            >
              <span
                class="symbol"
              >
                <test-file-stub
                  classname="root"
                />
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="root"
    >
      <div
        class="header"
        role="button"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-chevron-right fa-w-10 activeIndicator"
          data-icon="chevron-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
            fill="currentColor"
          />
        </svg>
        Advanced Options
      </div>
      <div
        class="body collapse"
      >
        <small
          class="text-muted font-italic"
        >
          No options to show
        </small>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
